keymanager/src/crypto/types.rs

```diff
@@ -96,13 +117,271 @@ pub struct InputKeyPair {
+    InvalidChecksum,
+    #[error("current epoch required")]
+    CurrentEpochRequired,
+    #[error("signature from the future")]
+    SignatureFromFuture,
+    #[error("signature expired")]
+    SignatureExpired,
+}
+
 /// Signed public key.
 #[derive(Clone, Debug, Default, PartialEq, Eq, cbor::Encode, cbor::Decode)]
 pub struct SignedPublicKey {
     /// Public key.
     pub key: PublicKey,
     /// Checksum of the key manager state.
     pub checksum: Vec<u8>,
-    /// Sign(sk, (key || checksum)) from the key manager.
+    /// Sign(sk, (key || checksum || runtime id || key pair id || epoch || expiration date) from
+    /// the key manager.
     pub signature: Signature,
 }
+
+impl SignedPublicKey {
+    /// Create a new signed public key.
+    pub fn new(
+        key: PublicKey,
+        checksum: Vec<u8>,
+        runtime_id: Namespace,
+        key_pair_id: KeyPairId,
+        epoch: Option<EpochTime>,
+        signer: &Arc<dyn Signer>,
+    ) -> Result<Self> {
+        if checksum.len() != CHECKSUM_SIZE {
+            return Err(SignedPublicKeyError::InvalidChecksum.into());
+        }
+
+        let body = Self::body(key, &checksum, runtime_id, key_pair_id, epoch);
+        let signature = signer.sign(&PUBLIC_KEY_CONTEXT, &body)?;
+
+        Ok(SignedPublicKey {
+            key,
+            checksum,
+            signature,
+        })
+    }
+
+    /// Verify the signature.
+    pub fn verify(
+        &self,
+        runtime_id: Namespace,
+        key_pair_id: KeyPairId,
+        epoch: Option<EpochTime>,
+        now: Option<EpochTime>,
+        pk: &EdPublicKey,
+    ) -> Result<()> {
+        // Checksum validation.
+        if self.checksum.len() != CHECKSUM_SIZE {
+            return Err(SignedPublicKeyError::InvalidChecksum.into());
+        }
+
+        // Cache validation for ephemeral keys.
+        if let Some(epoch) = epoch {
+            let now = now.ok_or(SignedPublicKeyError::CurrentEpochRequired)?;
+            if now < epoch {
+                return Err(SignedPublicKeyError::SignatureFromFuture.into());
+            }
+            if now > epoch + MAX_SIGNED_EPHEMERAL_PUBLIC_KEY_AGE {
+                return Err(SignedPublicKeyError::SignatureExpired.into());
+            }
+        }
+
+        let body = Self::body(self.key, &self.checksum, runtime_id, key_pair_id, epoch);
+        self.signature.verify(pk, &PUBLIC_KEY_CONTEXT, &body)
+    }
+
+    fn body(
+        key: PublicKey,
+        checksum: &[u8],
+        runtime_id: Namespace,
+        key_pair_id: KeyPairId,
+        epoch: Option<EpochTime>,
+    ) -> Vec<u8> {
+        let mut body = key.as_ref().to_vec();
+        body.extend_from_slice(checksum);
+        body.extend_from_slice(runtime_id.as_ref());
+        body.extend_from_slice(key_pair_id.as_ref());
+        if let Some(epoch) = epoch {
+            body.extend_from_slice(&epoch.to_be_bytes());
+            body.extend_from_slice(&MAX_SIGNED_EPHEMERAL_PUBLIC_KEY_AGE.to_be_bytes());
```

Comment:
This seems like it makes the scheme harder to upgrade if you want to actually change the value of the constant as old clients would deem responses from new runtimes invalid. Any downside to including the expiration field in the structure itself?

Reply:
Adding new field to the `SignedPublicKey` breaks old key manager clients as they will fail to decode cbor encoded RPC responses (extra field).
- We can make expiration field optional, disable cache expiration in the next core release and enable it after key manager clients are upgraded (probably in the next key manager release).
- We can set the expiration field only when public key request is made over an insecure channel (upgraded clients) and permanently enable after the next release.

The second solution would give us cache expiration immediately after the clients/runtimes upgrade.

Reply:
Yeah the second option sounds ok. Also speaking of versioning, should we introduce some way of versioning the protocol to make this easier in the future (could be done in a separate PR)?